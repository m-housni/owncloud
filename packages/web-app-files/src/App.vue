<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = eventBus.subscribe('drag-over', this.onDragOver)
    const dragOut = eventBus.subscribe('drag-out', this.hideDropzone)
    const drop = eventBus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
