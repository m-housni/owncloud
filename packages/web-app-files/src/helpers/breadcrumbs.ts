import { eventBus } from 'web-pkg/src/services/eventBus'
import { Location } from 'vue-router'
import omit from 'lodash-es/omit'

export interface BreadcrumbItem {
  text: string
  to?: Location
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentRoute: Location,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentRoute.path)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: {
          path: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/'),
          query: omit(currentRoute.query, 'fileId', 'page') // TODO: we need the correct fileId in the query. until we have that we must omit it because otherwise we would correct the path to the one of the (wrong) fileId.
        }
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => eventBus.publish('app.files.list.load')
    }
  ]
}
