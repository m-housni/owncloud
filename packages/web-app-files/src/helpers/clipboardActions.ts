export abstract class ClipboardActions {
  static readonly Cut = 'cut'
  static readonly Copy = 'copy'
}
