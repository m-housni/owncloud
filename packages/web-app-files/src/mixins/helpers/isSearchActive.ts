export default {
  computed: {
    $_isSearchActive(): boolean {
      return !!document.getElementById('files-global-search-options')
    }
  }
}
