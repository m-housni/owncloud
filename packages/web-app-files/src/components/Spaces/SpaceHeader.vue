<template>
  <div
    class="space-header oc-p-m"
    :class="{ 'oc-flex': !imageExpanded, 'space-header-squashed': sideBarOpen }"
  >
    <div class="space-header-image" :class="{ 'space-header-image-expanded': imageExpanded }">
      <img
        v-if="hasImage"
        class="oc-cursor-pointer"
        alt=""
        :src="imageContent"
        @click="toggleImageExpanded"
      />
      <div v-else class="space-header-image-default oc-flex oc-flex-middle oc-flex-center">
        <oc-icon name="layout-grid" size="xxlarge" class="oc-px-m oc-py-m" />
      </div>
    </div>
    <div class="space-header-infos">
      <div class="oc-flex oc-mb-s oc-flex-middle oc-flex-between">
        <div class="oc-flex oc-flex-middle space-header-infos-heading">
          <h1 class="space-header-name">{{ space.name }}</h1>
          <oc-button
            :id="`space-context-btn`"
            v-oc-tooltip="$gettext('Show context menu')"
            :aria-label="$gettext('Show context menu')"
            appearance="raw"
            class="oc-ml-s"
          >
            <oc-icon name="more-2" />
          </oc-button>
          <oc-drop
            :drop-id="`space-context-drop`"
            :toggle="`#space-context-btn`"
            mode="click"
            close-on-click
            :options="{ delayHide: 0 }"
            padding-size="small"
            position="right-start"
          >
            <space-context-actions :items="[space]" :space="space" />
          </oc-drop>
        </div>
        <oc-button
          v-if="memberCount"
          :aria-label="$gettext('Open context menu and show members')"
          appearance="raw"
          @click="openSideBarSharePanel"
        >
          <oc-icon name="group" fill-type="line" size="small" />
          <span class="space-header-people-count oc-text-small" v-text="memberCountString"></span>
        </oc-button>
      </div>
      <p v-if="space.description" class="oc-mt-rm oc-text-bold">{{ space.description }}</p>
      <div>
        <!-- eslint-disable vue/no-v-html -->
        <div ref="markdownContainerRef" class="markdown-container" v-html="markdownContent"></div>
        <!-- eslint-enable -->
        <div v-if="showMarkdownCollapse" class="markdown-collapse oc-text-center oc-mt-s">
          <oc-button appearance="raw" @click="toggleMarkdownCollapsed">
            <oc-icon :name="toggleMarkdownCollapsedIcon" />
            <span>{{ toggleMarkdownCollapsedText }}</span>
          </oc-button>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  onBeforeUnmount,
  onMounted,
  PropType,
  ref,
  unref,
  watch
} from 'vue'
import { SpaceResource } from 'web-client/src/helpers'
import { loadPreview } from 'web-pkg/src/helpers'
import {
  useAccessToken,
  useClientService,
  useStore,
  useTranslations
} from 'web-pkg/src/composables'
import { ImageDimension } from '../../constants'
import { configurationManager } from 'web-pkg/src/configuration'
import { VisibilityObserver } from 'web-pkg/src/observer'
import { marked } from 'marked'
import sanitizeHtml from 'sanitize-html'
import SpaceContextActions from './SpaceContextActions.vue'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { SideBarEventTopics } from '../../composables/sideBar'

const visibilityObserver = new VisibilityObserver()
const markdownContainerCollapsedClass = 'collapsed'

export default defineComponent({
  name: 'SpaceHeader',
  components: {
    SpaceContextActions
  },
  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: true
    },
    sideBarOpen: { type: Boolean, default: false }
  },
  setup(props) {
    const { $gettext, $ngettext, $gettextInterpolate } = useTranslations()
    const { getFileContents, getFileInfo } = useClientService().webdav
    const store = useStore()
    const userId = computed(() => store.getters.user?.id)
    const accessToken = useAccessToken({ store })

    const markdownContainerRef = ref(null)
    const markdownContent = ref('')
    const markdownCollapsed = ref(true)
    const showMarkdownCollapse = ref(false)
    const toggleMarkdownCollapsedIcon = computed(() => {
      return unref(markdownCollapsed) ? 'add' : 'subtract'
    })
    const toggleMarkdownCollapsedText = computed(() => {
      return unref(markdownCollapsed) ? $gettext('Show more') : $gettext('Show less')
    })
    const toggleMarkdownCollapsed = () => {
      markdownCollapsed.value = !unref(markdownCollapsed)
      unref(markdownContainerRef).classList.toggle(markdownContainerCollapsedClass)
    }
    const onMarkdownResize = () => {
      if (!unref(markdownContainerRef)) {
        return
      }

      unref(markdownContainerRef).classList.remove(markdownContainerCollapsedClass)
      const markdownContainerHeight = unref(markdownContainerRef).offsetHeight
      if (markdownContainerHeight < 150) {
        showMarkdownCollapse.value = false
        return
      }
      showMarkdownCollapse.value = true

      if (unref(markdownCollapsed)) {
        unref(markdownContainerRef).classList.add(markdownContainerCollapsedClass)
      }
    }
    const markdownResizeObserver = new ResizeObserver(onMarkdownResize)
    const observeMarkdownContainerResize = () => {
      if (!markdownResizeObserver || !unref(markdownContainerRef)) {
        return
      }
      markdownResizeObserver.unobserve(unref(markdownContainerRef))
      markdownResizeObserver.observe(unref(markdownContainerRef))
    }
    const unobserveMarkdownContainerResize = () => {
      if (!markdownResizeObserver || !unref(markdownContainerRef)) {
        return
      }
      markdownResizeObserver.unobserve(unref(markdownContainerRef))
    }
    onMounted(observeMarkdownContainerResize)
    onBeforeUnmount(() => {
      visibilityObserver.disconnect()
      unobserveMarkdownContainerResize()
    })
    watch(
      computed(() => props.space.spaceReadmeData),
      async (data: any) => {
        if (!data) {
          return
        }
        const decodedUri = decodeURI(data.webDavUrl)
        const webDavPathComponents = decodedUri.split('/')
        const idComponent = webDavPathComponents.find((c) => c.startsWith(props.space.id as string))
        if (!idComponent) {
          return
        }
        const path = webDavPathComponents
          .slice(webDavPathComponents.indexOf(idComponent) + 1)
          .join('/')

        const fileContentsResponse = await getFileContents(props.space, {
          path: decodeURIComponent(path)
        })

        unobserveMarkdownContainerResize()
        const parsedMarkdown = marked.parse(fileContentsResponse.body)
        // Sanitize markdown content to prevent XSS vulnerabilities
        markdownContent.value = sanitizeHtml(parsedMarkdown)

        if (unref(markdownContent)) {
          observeMarkdownContainerResize()
        }
      },
      { deep: true, immediate: true }
    )

    const imageContent = ref(null)
    const imageExpanded = ref(false)
    const toggleImageExpanded = () => {
      imageExpanded.value = !unref(imageExpanded)
    }
    watch(
      computed(() => props.space.spaceImageData),
      async (data) => {
        if (!data) {
          return
        }
        const decodedUri = decodeURI(props.space.spaceImageData.webDavUrl)
        const webDavPathComponents = decodedUri.split('/')
        const idComponent = webDavPathComponents.find((c) => c.startsWith(props.space.id as string))
        if (!idComponent) {
          return
        }
        const path = webDavPathComponents
          .slice(webDavPathComponents.indexOf(idComponent) + 1)
          .join('/')

        const resource = await getFileInfo(props.space, {
          path: decodeURIComponent(path)
        })
        imageContent.value = await loadPreview({
          resource,
          isPublic: false,
          dimensions: ImageDimension.Preview,
          server: configurationManager.serverUrl,
          userId: unref(userId),
          token: unref(accessToken)
        })
      },
      { deep: true, immediate: true }
    )
    const hasImage = computed(() => {
      return props.space?.spaceImageData
    })

    const memberCount = computed(() => {
      return store.getters['runtime/spaces/spaceMembers'].length
    })
    const memberCountString = computed(() => {
      const translated = $ngettext('%{count} member', '%{count} members', unref(memberCount))

      return $gettextInterpolate(translated, {
        count: unref(memberCount)
      })
    })

    const openSideBarSharePanel = () => {
      store.commit('Files/SET_SELECTED_IDS', [])
      eventBus.publish(SideBarEventTopics.openWithPanel, 'space-share')
    }

    return {
      markdownContainerRef,
      markdownContent,
      markdownCollapsed,
      showMarkdownCollapse,
      toggleMarkdownCollapsedIcon,
      toggleMarkdownCollapsedText,
      toggleMarkdownCollapsed,
      imageContent,
      imageExpanded,
      toggleImageExpanded,
      hasImage,
      memberCount,
      memberCountString,
      openSideBarSharePanel
    }
  }
})
</script>

<style lang="scss">
.space-header {
  &-squashed {
    .space-header-image {
      @media only screen and (max-width: 1200px) {
        display: none;
      }
    }
  }

  &-image {
    width: 280px;
    min-width: 280px;
    aspect-ratio: 16 / 9;
    margin-right: var(--oc-space-large);
    max-height: 158px;
    &-default {
      background-color: var(--oc-color-background-highlight);
      height: 100%;
      border-radius: 10px;
    }
    &-expanded {
      width: 100%;
      margin: 0;
      max-height: 100%;
      max-width: 100%;
    }
    img {
      border-radius: 10px;
      width: 100%;
      max-height: 100%;
      object-fit: cover;
    }
  }

  &-infos {
    flex: 1;
    &-heading {
      max-width: 100%;
    }
  }

  &-name {
    font-size: 1.5rem;
    word-break: break-all;
  }

  &-people-count {
    white-space: nowrap;
  }

  .markdown-container.collapsed {
    max-height: 100px;
    overflow: hidden;
    -webkit-mask-image: linear-gradient(180deg, #000 90%, transparent);
  }
}
</style>
