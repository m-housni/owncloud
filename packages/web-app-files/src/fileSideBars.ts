import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import SharesPanel from './components/SideBar/Shares/SharesPanel.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from './components/SideBar/Details/SpaceDetails.vue'
import {
  isLocationTrashActive,
  isLocationPublicActive,
  isLocationSpacesActive,
  isLocationSharesActive
} from './router'
import { spaceRoleEditor, spaceRoleManager } from 'web-client/src/helpers/share'

import { Panel } from '../../web-pkg/src/components/sideBar'

import { Resource, User } from 'web-client'
import Router from 'vue-router'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  highlightedFile,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  highlightedFile: Resource
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ rootFolder, highlightedFile }): Panel => ({
    app: 'no-selection',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return !highlightedFile || (rootFolder && highlightedFile?.type !== 'space')
    }
  }),
  ({ router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'details',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default: !isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return (
        !isLocationTrashActive(router, 'files-trash-generic') &&
        !multipleSelection &&
        !rootFolder &&
        highlightedFile
      )
    }
  }),
  ({ multipleSelection, rootFolder, highlightedFile, router }) => ({
    app: 'details-multiple',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    componentAttrs: {
      get showSpaceCount() {
        return (
          !isLocationSpacesActive(router, 'files-spaces-generic') &&
          !isLocationSharesActive(router, 'files-shares-with-me') &&
          !isLocationTrashActive(router, 'files-trash-generic')
        )
      }
    },
    default: () => true,
    get enabled() {
      return multipleSelection && (!rootFolder || highlightedFile?.type === 'space')
    }
  }),
  ({ multipleSelection, highlightedFile }) => ({
    app: 'details-space',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: () => true,
    get enabled() {
      return highlightedFile?.type === 'space' && !multipleSelection
    }
  }),
  ({ router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default: isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return !multipleSelection && !rootFolder && highlightedFile
    }
  }),
  ({ multipleSelection, highlightedFile, user }) => ({
    app: 'space-actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (multipleSelection) {
        return false
      }
      if (highlightedFile?.type !== 'space') {
        return false
      }
      return [
        ...highlightedFile.spaceRoles[spaceRoleManager.name],
        ...highlightedFile.spaceRoles[spaceRoleEditor.name]
      ].includes(user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'sharing',
    icon: 'user-add',
    iconFillType: 'line',
    title: $gettext('Shares'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: false,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      if (multipleSelection || rootFolder || !highlightedFile) {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ multipleSelection, highlightedFile, capabilities }) => ({
    app: 'space-share',
    icon: 'group',
    title: $gettext('Members'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: true,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      return highlightedFile?.type === 'space' && !multipleSelection
    }
  }),
  ({ capabilities, highlightedFile, router, multipleSelection, rootFolder }) => ({
    app: 'versions',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }
      return !!capabilities.core && highlightedFile && highlightedFile.type !== 'folder'
    }
  })
]

export default panelGenerators
