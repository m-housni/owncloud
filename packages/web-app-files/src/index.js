import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { archiverService, thumbnailService, Registry } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = [
  {
    name(capabilities) {
      return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
    },
    icon: appInfo.icon,
    route: {
      path: `/${appInfo.id}/spaces/personal`
    }
  },
  {
    name: $gettext('Favorites'),
    icon: 'star',
    route: {
      path: `/${appInfo.id}/favorites`
    },
    enabled(capabilities) {
      return capabilities.files && capabilities.files.favorites
    }
  },
  {
    name: $gettext('Shares'),
    icon: 'share-forward',
    route: {
      path: `/${appInfo.id}/shares`
    },
    activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
    enabled(capabilities) {
      return capabilities.files_sharing?.api_enabled !== false
    }
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appInfo.id}/spaces/projects`
    },
    activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
    enabled(capabilities) {
      return capabilities.spaces && capabilities.spaces.projects === true
    }
  },
  {
    name: $gettext('Deleted files'),
    icon: 'delete-bin-5',
    route: {
      path: `/${appInfo.id}/trash`
    },
    enabled(capabilities) {
      return capabilities.dav && capabilities.dav.trashbin === '1.0'
    }
  }
]

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store }) {
    Registry.sdkSearch = new SDKSearch(store, router)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)

    archiverService.initialize(
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
    // FIXME: Use capability data only as soon as available
    thumbnailService.initialize(
      get(store, 'getters.capabilities.files.thumbnail', {
        enabled: true,
        version: 'v0.1',
        supportedMimeTypes: store.getters.configuration?.options?.previewFileMimeTypes || []
      })
    )
  }
}
