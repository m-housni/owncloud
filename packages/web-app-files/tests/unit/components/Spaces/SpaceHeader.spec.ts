import Vuex from 'vuex'
import { mount } from '@vue/test-utils'
import SpaceHeader from 'web-app-files/src/components/Spaces/SpaceHeader.vue'
import { createStore } from 'vuex-extensions'
import { defaultComponentMocks } from 'web-test-helpers/src/mocks/defaultComponentMocks'
import { buildSpace } from 'web-client/src/helpers'
import { defaultLocalVue } from 'web-test-helpers/src/localVue/defaultLocalVue'
import { defaultStoreMockOptions } from 'web-test-helpers/src/mocks/store/defaultStoreMockOptions'

window.ResizeObserver =
  window.ResizeObserver ||
  jest.fn().mockImplementation(() => ({
    disconnect: jest.fn(),
    observe: jest.fn(),
    unobserve: jest.fn()
  }))

afterEach(() => jest.clearAllMocks())

describe('SpaceHeader', () => {
  it('should add the "squashed"-class when the sidebar is opened', () => {
    const wrapper = getWrapper({ space: buildSpace({ id: 1 }), sideBarOpen: true })
    expect(wrapper.find('.space-header-squashed').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  describe('space image', () => {
    it('should show the default image if no other image is set', () => {
      const wrapper = getWrapper({ space: buildSpace({ id: 1 }) })
      expect(wrapper.find('.space-header-image-default').exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
    it('should show the set image', () => {
      const spaceMock = { spaceImageData: { webDavUrl: '/' } }
      const wrapper = getWrapper({ space: { ...buildSpace({ id: 1 }), ...spaceMock } })
      expect(wrapper.find('.space-header-image-default').exists()).toBeFalsy()
      expect(wrapper.find('.space-header-image img').exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function getWrapper({ space = {}, sideBarOpen = false }) {
  const localVue = defaultLocalVue()
  const mocks = {
    ...defaultComponentMocks()
  }
  const storeOptions = {
    ...defaultStoreMockOptions,
    modules: {
      ...defaultStoreMockOptions.modules,
      runtime: {
        namespaced: true,
        modules: {
          spaces: {
            namespaced: true,
            getters: {
              spaceMembers: () => []
            }
          }
        }
      }
    }
  }
  const store = createStore(Vuex.Store, storeOptions)
  return mount(SpaceHeader, {
    localVue,
    mocks,
    store,
    propsData: {
      space,
      sideBarOpen
    }
  })
}
