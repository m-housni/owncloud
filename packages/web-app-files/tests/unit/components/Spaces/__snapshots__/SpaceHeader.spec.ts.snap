// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceHeader should add the "squashed"-class when the sidebar is opened 1`] = `
<div class="space-header oc-p-m oc-flex space-header-squashed">
  <div class="space-header-image">
    <div class="space-header-image-default oc-flex oc-flex-middle oc-flex-center"><span class="oc-px-m oc-py-m oc-icon oc-icon-xxl oc-icon-passive"><!----></span></div>
  </div>
  <div class="space-header-infos">
    <div class="oc-flex oc-mb-s oc-flex-middle oc-flex-between">
      <div class="oc-flex oc-flex-middle space-header-infos-heading">
        <h1 class="space-header-name"></h1> <button aria-label="Show context menu" type="button" class="oc-ml-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="space-context-btn"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
        <div id="space-context-drop" class="oc-drop oc-box-shadow-medium oc-rounded" options="[object Object]">
          <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
            <div>
              <div id="oc-files-context-menu">
                <ul id="oc-files-context-actions-members" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-files-context-actions-border">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Members</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
                <ul id="oc-files-context-actions-trashBin" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pt-s">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-delete-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Deleted files</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
              </div>
              <!---->
              <!----> <input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
    <!---->
    <div>
      <div class="markdown-container"></div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`SpaceHeader space image should show the default image if no other image is set 1`] = `
<div class="space-header oc-p-m oc-flex">
  <div class="space-header-image">
    <div class="space-header-image-default oc-flex oc-flex-middle oc-flex-center"><span class="oc-px-m oc-py-m oc-icon oc-icon-xxl oc-icon-passive"><!----></span></div>
  </div>
  <div class="space-header-infos">
    <div class="oc-flex oc-mb-s oc-flex-middle oc-flex-between">
      <div class="oc-flex oc-flex-middle space-header-infos-heading">
        <h1 class="space-header-name"></h1> <button aria-label="Show context menu" type="button" class="oc-ml-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="space-context-btn"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
        <div id="space-context-drop" class="oc-drop oc-box-shadow-medium oc-rounded" options="[object Object]">
          <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
            <div>
              <div id="oc-files-context-menu">
                <ul id="oc-files-context-actions-members" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-files-context-actions-border">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Members</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
                <ul id="oc-files-context-actions-trashBin" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pt-s">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-delete-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Deleted files</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
              </div>
              <!---->
              <!----> <input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
    <!---->
    <div>
      <div class="markdown-container"></div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`SpaceHeader space image should show the set image 1`] = `
<div class="space-header oc-p-m oc-flex">
  <div class="space-header-image"><img alt="" class="oc-cursor-pointer"></div>
  <div class="space-header-infos">
    <div class="oc-flex oc-mb-s oc-flex-middle oc-flex-between">
      <div class="oc-flex oc-flex-middle space-header-infos-heading">
        <h1 class="space-header-name"></h1> <button aria-label="Show context menu" type="button" class="oc-ml-s oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="space-context-btn"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
        <div id="space-context-drop" class="oc-drop oc-box-shadow-medium oc-rounded" options="[object Object]">
          <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
            <div>
              <div id="oc-files-context-menu">
                <ul id="oc-files-context-actions-members" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-files-context-actions-border">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Members</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
                <ul id="oc-files-context-actions-trashBin" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pt-s">
                  <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-delete-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Deleted files</span>
                      <!---->
                      <!---->
                    </button></li>
                </ul>
              </div>
              <!---->
              <!----> <input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
    <!---->
    <div>
      <div class="markdown-container"></div>
      <!---->
    </div>
  </div>
</div>
`;
